File: examples/quarter_truck/quarter_truck_ssp.cpp
#include "ecos/algorithm/fixed_step_algorithm.hpp"
#include "ecos/listeners/csv_writer.hpp"
#include "ecos/logger/logger.hpp"
#include "ecos/ssp/ssp_loader.hpp"
#include "ecos/util/plotter.hpp"

#include <spdlog/stopwatch.h>

using namespace ecos;

int main()
{
    log::set_logging_level(ecos::log::level::debug);

    const std::filesystem::path sspFolder = std::string(DATA_FOLDER) + "/ssp/quarter_truck";

    try {
        const auto ss = load_ssp(sspFolder);
        const auto sim = ss->load(std::make_unique<fixed_step_algorithm>(1.0 / 100));

        auto csvWriter = std::make_unique<csv_writer>("results/quarter_truck.csv");
        const auto outputPath = csvWriter->output_path();
        csv_config& config = csvWriter->config();
        config.register_variable("chassis::zChassis");
        config.register_variable("wheel::zWheel");
        config.register_variable("ground::zGround");
        sim->add_listener("csv_writer", std::move(csvWriter));

        spdlog::stopwatch sw;
        sim->init("initialValues");
        sim->step_until(5);
        log::info("Elapsed {:.4f}s", sw);

        sim->terminate();

        plot_csv(outputPath, sspFolder / "ChartConfig.xml");
    } catch (const std::exception& ex) {

        log::err(ex.what());
    }
}

File: examples/quarter_truck/quarter_truck_ssp.py
from ecospy import *
from ecospy.plotter import *

from pathlib import Path


def main():
    print(f"Ecoslib version: {EcosLib().version()}")

    EcosLib().set_log_level("debug")

    ssp_dir = (Path(__file__).parent.parent.parent / 'data' / 'ssp' / '1.0' / 'quarter_truck').resolve()
    ssp_file = f"{ssp_dir}/quarter-truck.ssp"
    result_file = "results/python/quarter_truck_ssp.csv"

    with(EcosSimulation(ssp_path=ssp_file, step_size=1.0 / 100)) as sim:
        sim.add_csv_writer(result_file, f"{ssp_dir}/CsvConfig.xml")

        sim.init(parameter_set="initialValues")
        sim.step_until(time_point=10)
        sim.terminate()

    config = TimeSeriesConfig(
        title="Quarter-truck",
        y_label="Height[m]",
        identifiers=["chassis::zChassis", "wheel::zWheel", "ground::zGround"])
    plotter = Plotter(result_file, config)
    plotter.show()


if __name__ == "__main__":
    import sys

    if sys.platform != "win32":
        raise EnvironmentError("This particular example (quarter_truck) can only be run on Windows.")

    main()
